### PR TITLE
feat(router): add support for nested routers

### LINKED ISSUE #6: Sub-routers
The idea is to be able to use a router as path of an another router. Maybe this is supported, if so, sorry for creating a duplicate issue.

### PR DESCRIPTION
**Pull request by c4spar**
Closes #6 

This implements support for nested routers:

```typescript
const posts = new Router()
  .get("/", (ctx) => {
    ctx.response.body = `Forum: ${ctx.params.forumId}`;
  })
  .get("/:postId", (ctx) => {
    ctx.response.body =
      `Forum: ${ctx.params.forumId}, Post: ${ctx.params.postId}`;
  });

const forums = new Router()
  .get("/forums/:forumId/posts", posts.routes(), posts.allowedMethods());

// Responds to "http://localhost:8000/forums/oak/posts" and "http://localhost:8000/forums/oak/posts/nested-routers"
await new Application()
  .use(forums.routes())
  .listen({ port: 8000 });
```

### CODE REVIEW

**Comment by mmintel** [comment]
hi there! great to see that somebody is working on subrouting :) just from reading your documentation: I find it confusing that you nest the router in a `.get`, what about the other methods like `.post` and such?

**Comment by c4spar** [comment]
It works also with all other methods like `.post` and `.use`.
I just copied the example from the koa readme. I can add a better documentation but i am waiting for feedback if the PR is welcome or if there is already a different solution planed.

**Comment by ewired** [comment]
Thanks; I can confirm that this patch is working as expected in my application via `router.use("/path", anotherRouter.routes())`. LGTM

For the time being, don't try this at home:
```ts
export {
  Application,
  Router,
  Status,
} from "https://denopkg.com/c4spar/oak@60465d37b61e5b820c1e3f6deb3d8221cb5fc9db/mod.ts";
```

File: router.ts [hunk 1]
```diff
@@ -371,19 +381,62 @@ export class Router<
...
       for (const p of path) {
-        this.#register(p, middleware, methods, options);
+        this.#register(p, middlewares, methods, options);
       }
       return;
     }
 
+    let layerMiddlewares = [];
+    for (let i = 0; i < middlewares.length; i++) {
+      const middleware = middlewares[i];
```
Comment (Husenap) [inline router.ts:397]:
```suggestion
    for (const middleware of middlewares) {
```

**Review by Husenap — CHANGES_REQUESTED** [review]
I'm just nitpicking here, but personally I would rewrite this loop to be consistent with the other loops in the surrounding code and also because I don't see the index being used.

Other than that I think it looks good 👍

**Comment by c4spar** [comment]
Hey @kitsonk, what du you think about this PR?

It implements support for nested routers in a similar way as koa it does, by:
* cloning the router and its layers
* updating the prefix of the cloned layers
* and adding the cloned layers to the main router

The router is passed via the middleware to the parent router, which is returned by the `.routes()` method. This was already implemented and i only exposed the `.router` property on the `Middleware` interface.

The only difference to koa is, that all methods can mount a router, not only the `.use()` method. If this is not wanted, i can move the logic to the `.use()` method, so that it works the same way as with koa.

---

Some years ago, i created also a routing module in nodejs that was inspired by koa, but i experimented with a different approach to mount nested routers.

Instead of passing the router to it's parent, cloning it and appinding its layers to the main router, i added a second midleware type `MountMiddleware` which takes the prefix of all parent routers as third argument.
The mount middleware is also passed to the parent router with the `.routes()` method and has a property `middleware.mountable` to be able to distinguish them from a normal middleware and to pass the prefix only to the mount middlewares.

The difference of this solutions is, that the real router instance will be mounted instead of a cloned one.
If we clone the router, we will loose the reference to the router which makes it impossible to make changes on the router after it was mounted.

For example, this is not possible with the first solution, but it would work with this solution:

```typescript
const router1 = new Router();
const router2 = new Router();

router1.use("/foo/bar", router2.routes());

router2.get("/:baz", async () => {
  // this middleware will not be mounted on router1 because it was added after router2 was mounted on router1
});
```

Some pros and cons:

**pros**

* it's possible to add/remove middlewares/routers after mounting and at runtime.
* we don't need to clone the router which leads to a borken reference to the real router instance.
* a bit faster startup time

**cons**

* maybe a bit slower request time because we need to create the path regex during the request to create it with the correct parent prefix, but this could be solved with caching

---

Right now, i was just going with the first solution because it was easier to implement and i wasn't sure because of the performnce. But i have created a new repository where i'm experimenting with this solution.
If you are interrested for testing this solution with oak, i can try to implement it in a second branch.
Here is my repository where i'm experimenting with this solution: https://github.com/c4spar/deno-deploy-route

**Comment by kitsonk** [comment]
@c4spar sorry it took me so long to look at this... thanks for the contribution... I made some minor edits to the code, let me know if you are ok with them and I will merge the PR.

**Comment by c4spar** [comment]
hey @kitsonk, i know you have a lot to do with deno and it's not a small change, so no problem! 😉
I'm fine with your changes! Thanks! 🙂

**Review by kitsonk — APPROVED** [review]
LGTM, thank you for the contribution and being patient with getting it landed.

**Pull request merged**